Source/Core/VideoCommon/NetPlayChatUI.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright 2019 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "VideoCommon/NetPlayChatUI.h"
+
+#include <imgui.h>
+
+constexpr float DEFAULT_WINDOW_WIDTH = 220.0f;
+constexpr float DEFAULT_WINDOW_HEIGHT = 400.0f;
+
+constexpr size_t MAX_BACKLOG_SIZE = 100;
+
+std::unique_ptr<NetPlayChatUI> g_netplay_chat_ui;
+
+NetPlayChatUI::NetPlayChatUI(std::function<void(const std::string&)> callback)
+{
+  m_message_callback = std::move(callback);
+}
+
+void NetPlayChatUI::Display()
+{
+  const float scale = ImGui::GetIO().DisplayFramebufferScale.x;
+
+  ImGui::SetNextWindowPos(ImVec2(10.0f * scale, 10.0f * scale), ImGuiCond_FirstUseEver);
+  ImGui::SetNextWindowSizeConstraints(
+      ImVec2(DEFAULT_WINDOW_WIDTH * scale, DEFAULT_WINDOW_HEIGHT * scale),
+      ImGui::GetIO().DisplaySize);
+
+  if (!ImGui::Begin("Chat", nullptr, ImGuiWindowFlags_None))
+  {
+    ImGui::End();
+    return;
+  }
+
+  ImGui::BeginChild("Scrolling", ImVec2(0, -30 * scale), true, ImGuiWindowFlags_None);
+  for (const auto& msg : m_messages)
+  {
+    auto c = msg.second;
+    ImGui::PushTextWrapPos(0.0f);
+    ImGui::TextColored(ImVec4(c[0], c[1], c[2], 1.0f), "%s", msg.first.c_str());
+    ImGui::PopTextWrapPos();
+  }
+
+  if (m_scroll_to_bottom)
+  {
+    ImGui::SetScrollHere(1.0f);
+    m_scroll_to_bottom = false;
+  }
+
+  m_is_scrolled_to_bottom = ImGui::GetScrollY() == ImGui::GetScrollMaxY();
+
+  ImGui::EndChild();
+
+  ImGui::Spacing();
+
+  ImGui::PushItemWidth(-50.0f * scale);
+
+  if (ImGui::InputText("", m_message_buf, IM_ARRAYSIZE(m_message_buf),
+                       ImGuiInputTextFlags_EnterReturnsTrue))
+  {
+    SendMessage();
+    ImGui::SetKeyboardFocusHere(-1);
+  }
+
+  ImGui::PopItemWidth();
+
+  ImGui::SameLine();
+
+  if (ImGui::Button("Send"))
+    SendMessage();
+
+  ImGui::End();
+}
+
+void NetPlayChatUI::AppendChat(const std::string& message, const std::string& color)
+{
+  if (m_messages.size() > MAX_BACKLOG_SIZE)
+    m_messages.pop_front();
+
+  std::array<float, 3> c;
+
+  if (color == "magenta")
```

Comment:
Having four hard-coded color names is odd. Any reason we can't have some constants and pass the actual colors to `AppendChat`?

Reply:
+1 for an enum + lookup table, or just storing ImVec4s.

Reply:
I was about to suggest something similar the other day, but then noticed the very same parameter/string is used directly in the Qt interface (where it just ends up in some QML Markup or whatever to produce text of that color) - with both color names and `#rrggbb`
Would take away some flexibility; not sure if we need that tho.